File: python/mirage/commands/builtin/databricks_volume/ls.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from mirage.accessor.databricks_volume import DatabricksVolumeAccessor
from mirage.cache.index import IndexCacheStore
from mirage.commands.builtin.utils.formatting import _human_size
from mirage.commands.registry import command
from mirage.commands.spec import SPECS
from mirage.core.databricks_volume.glob import resolve_glob
from mirage.core.databricks_volume.readdir import readdir
from mirage.core.databricks_volume.stat import stat
from mirage.io.types import ByteSource, IOResult
from mirage.types import FileStat, FileType, PathSpec


async def _ls_entries(
    accessor: DatabricksVolumeAccessor,
    path: PathSpec,
    all_files: bool,
    sort_by: str,
    reverse: bool,
    recursive: bool,
    list_dir: bool,
    warnings: list[str],
    index: IndexCacheStore | None = None,
) -> list[FileStat]:
    if list_dir:
        return [await stat(accessor, path, index)]

    try:
        entries = await readdir(accessor, path, index)
    except (FileNotFoundError, ValueError) as exc:
        warnings.append(f"ls: cannot access '{path.original}': {exc}")
        return []

    stats: list[FileStat] = []
    for entry in entries:
        try:
            entry_spec = PathSpec(
                original=entry,
                directory=entry,
                resolved=False,
                prefix=path.prefix,
            )
            entry_stat = await stat(accessor, entry_spec, index)
        except (FileNotFoundError, ValueError) as exc:
            warnings.append(f"ls: cannot access '{entry}': {exc}")
            continue
        if not all_files and entry_stat.name.startswith("."):
            continue
        stats.append(entry_stat)

    if sort_by == "time":
        stats.sort(key=lambda entry: entry.modified or "", reverse=not reverse)
    elif sort_by == "size":
        stats.sort(key=lambda entry: entry.size or 0, reverse=not reverse)
    else:
        stats.sort(key=lambda entry: entry.name, reverse=reverse)

    if recursive:
        sub_entries: list[FileStat] = []
        for entry_stat in stats:
            sub_entries.append(entry_stat)
            if entry_stat.type == FileType.DIRECTORY:
                entry_path = path.child(entry_stat.name)
                entry_spec = PathSpec(
                    original=entry_path,
                    directory=entry_path,
                    resolved=False,
                    prefix=path.prefix,
                )
                sub_entries.extend(await _ls_entries(
                    accessor,
                    entry_spec,
                    all_files,
                    sort_by,
                    reverse,
                    recursive,
                    False,
                    warnings,
                    index,
                ))
        return sub_entries

    return stats


@command("ls", resource="databricks_volume", spec=SPECS["ls"])
async def ls(
    accessor: DatabricksVolumeAccessor,
    paths: list[PathSpec],
    *texts: str,
    stdin: bytes | None = None,
    args_l: bool = False,
    args_1: bool = False,
    a: bool = False,
    A: bool = False,
    h: bool = False,
    t: bool = False,
    S: bool = False,
    r: bool = False,
    R: bool = False,
    d: bool = False,
    F: bool = False,
    index: IndexCacheStore = None,
    cwd: PathSpec | str = "/",
    prefix: str = "",
    **_extra: object,
) -> tuple[ByteSource | None, IOResult]:
    if not paths:
        cwd_str = cwd.original if isinstance(cwd, PathSpec) else cwd
        cwd_prefix = cwd.prefix if isinstance(cwd, PathSpec) else ""
        paths = [
            PathSpec(
                original=cwd_str,
                directory=cwd_str,
                resolved=False,
                prefix=cwd_prefix,
            )
        ]
    paths = await resolve_glob(accessor, paths, index)
    all_files = a or A
    sort_by = "name"
    if t:
        sort_by = "time"
    elif S:
        sort_by = "size"
    warnings: list[str] = []
    results: list[str] = []
    for path in paths:
        try:
            entries = await _ls_entries(
                accessor,
                path,
                all_files,
                sort_by,
                r,
                R,
                d,
                warnings,
                index,
            )
        except (FileNotFoundError, ValueError) as exc:
            warnings.append(f"ls: cannot access '{path.original}': {exc}")
            continue
        if args_l and not args_1:
            for entry in entries:
                size_str = _human_size(entry.size or 0) if h else str(
                    entry.size or 0)
                results.append(f"{entry.type or '-'}\t{size_str}\t"
                               f"{entry.modified or ''}\t{entry.name}")
        else:
            for entry in entries:
                is_dir = F and entry.type == FileType.DIRECTORY
                results.append(entry.name + ("/" if is_dir else ""))
    stderr = "\n".join(warnings).encode() if warnings else None
    exit_code = 1 if warnings and not results else 0
    return "\n".join(results).encode(), IOResult(
        stderr=stderr,
        exit_code=exit_code,
    )

File: typescript/packages/node/src/commands/builtin/ssh/ls/ls.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import {
  FileType,
  IOResult,
  PathSpec,
  ResourceName,
  command,
  specOf,
  type ByteSource,
  type CommandFnResult,
  type CommandOpts,
  type FileStat,
  rstripSlash,
} from '@struktoai/mirage-core'
import { stat as sshStat } from '../../../../core/ssh/stat.ts'
import { readdir as sshReaddir } from '../../../../core/ssh/readdir.ts'
import type { SSHAccessor } from '../../../../accessor/ssh.ts'

function humanSize(n: number): string {
  const units = ['B', 'K', 'M', 'G', 'T']
  let v = n
  let i = 0
  while (v >= 1024 && i < units.length - 1) {
    v /= 1024
    i += 1
  }
  const s = v >= 10 || i === 0 ? Math.round(v).toString() : v.toFixed(1)
  return `${s}${units[i] ?? ''}`
}

function childSpec(entryPath: string, prefix: string): PathSpec {
  return new PathSpec({
    original: entryPath,
    directory: entryPath,
    resolved: false,
    prefix,
  })
}

function formatEntry(s: FileStat, long: boolean, human: boolean, classify: boolean): string {
  if (long) {
    const size = human ? humanSize(s.size ?? 0) : String(s.size ?? 0)
    return `${s.type ?? '-'}\t${size}\t${s.modified ?? ''}\t${s.name}`
  }
  const suffix = classify && s.type === FileType.DIRECTORY ? '/' : ''
  return `${s.name}${suffix}`
}

function sortStats(
  stats: FileStat[],
  sortBy: 'time' | 'size' | 'name',
  reverse: boolean,
): FileStat[] {
  const sorted = [...stats].sort((a, b) => {
    if (sortBy === 'time') return (b.modified ?? '').localeCompare(a.modified ?? '')
    if (sortBy === 'size') return (b.size ?? 0) - (a.size ?? 0)
    return a.name.localeCompare(b.name)
  })
  if (reverse) sorted.reverse()
  return sorted
}

async function listDir(accessor: SSHAccessor, dir: PathSpec, all: boolean): Promise<FileStat[]> {
  const entries = await sshReaddir(accessor, dir)
  const stats = await Promise.all(entries.map((p) => sshStat(accessor, childSpec(p, dir.prefix))))
  return all ? stats : stats.filter((s) => !s.name.startsWith('.'))
}

interface WalkOpts {
  all: boolean
  long: boolean
  human: boolean
  classify: boolean
  sortBy: 'time' | 'size' | 'name'
  reverse: boolean
}

async function walkRecursive(
  accessor: SSHAccessor,
  dir: PathSpec,
  opts: WalkOpts,
  header: boolean,
  lines: string[],
  warnings: string[],
): Promise<void> {
  let stats: FileStat[]
  try {
    stats = await listDir(accessor, dir, opts.all)
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err)
    warnings.push(`ls: cannot access '${dir.original}': ${msg}`)
    return
  }
  if (header) lines.push(`${dir.stripPrefix}:`)
  const sorted = sortStats(stats, opts.sortBy, opts.reverse)
  for (const s of sorted) lines.push(formatEntry(s, opts.long, opts.human, opts.classify))
  const subdirs = sorted.filter((s) => s.type === FileType.DIRECTORY)
  for (const sub of subdirs) {
    lines.push('')
    const base = rstripSlash(dir.stripPrefix)
    const childPath = `${base}/${sub.name}`
    await walkRecursive(accessor, childSpec(childPath, dir.prefix), opts, true, lines, warnings)
  }
}

async function lsCommand(
  accessor: SSHAccessor,
  paths: PathSpec[],
  _texts: string[],
  opts: CommandOpts,
): Promise<CommandFnResult> {
  const targets: PathSpec[] =
    paths.length > 0
      ? paths
      : [
          new PathSpec({
            original: opts.cwd,
            directory: opts.cwd,
            resolved: false,
            prefix: opts.mountPrefix ?? '',
          }),
        ]
  const long = opts.flags.args_l === true && opts.flags.args_1 !== true
  const all = opts.flags.a === true || opts.flags.A === true
  const human = opts.flags.h === true
  const reverse = opts.flags.r === true
  const classify = opts.flags.F === true
  const recursive = opts.flags.R === true
  const listDirItself = opts.flags.d === true
  const sortBy: 'time' | 'size' | 'name' =
    opts.flags.t === true ? 'time' : opts.flags.S === true ? 'size' : 'name'
  const warnings: string[] = []
  const lines: string[] = []

  if (listDirItself) {
    for (const p of targets) {
      try {
        const s = await sshStat(accessor, p)
        lines.push(formatEntry(s, long, human, classify))
      } catch (err) {
        const msg = err instanceof Error ? err.message : String(err)
        warnings.push(`ls: cannot access '${p.original}': ${msg}`)
      }
    }
    const out: ByteSource = new TextEncoder().encode(lines.join('\n'))
    const exitCode = warnings.length > 0 && lines.length === 0 ? 1 : 0
    if (warnings.length > 0) {
      const stderr = new TextEncoder().encode(warnings.join('\n'))
      return [out, new IOResult({ stderr, exitCode })]
    }
    return [out, new IOResult({ exitCode })]
  }

  if (recursive) {
    const walkOpts: WalkOpts = { all, long, human, classify, sortBy, reverse }
    for (let i = 0; i < targets.length; i++) {
      const p = targets[i]
      if (p === undefined) continue
      if (i > 0) lines.push('')
      await walkRecursive(accessor, p, walkOpts, true, lines, warnings)
    }
    const out: ByteSource = new TextEncoder().encode(lines.join('\n'))
    const exitCode = warnings.length > 0 && lines.length === 0 ? 1 : 0
    if (warnings.length > 0) {
      const stderr = new TextEncoder().encode(warnings.join('\n'))
      return [out, new IOResult({ stderr, exitCode })]
    }
    return [out, new IOResult({ exitCode })]
  }

  for (const p of targets) {
    let stats: FileStat[]
    try {
      stats = await listDir(accessor, p, all)
    } catch (err) {
      try {
        stats = [await sshStat(accessor, p)]
      } catch {
        const msg = err instanceof Error ? err.message : String(err)
        warnings.push(`ls: cannot access '${p.original}': ${msg}`)
        continue
      }
    }
    for (const s of sortStats(stats, sortBy, reverse))
      lines.push(formatEntry(s, long, human, classify))
  }
  const out: ByteSource = new TextEncoder().encode(lines.join('\n'))
  const exitCode = warnings.length > 0 && lines.length === 0 ? 1 : 0
  if (warnings.length > 0) {
    const stderr = new TextEncoder().encode(warnings.join('\n'))
    return [out, new IOResult({ stderr, exitCode })]
  }
  return [out, new IOResult({ exitCode })]
}

export const SSH_LS = command({
  name: 'ls',
  resource: ResourceName.SSH,
  spec: specOf('ls'),
  fn: lsCommand,
})
